src/design_notes/copy_ergonomics.md

```diff
@@ -0,0 +1,59 @@
+# Copy type ergonomics
+
+## Background
+
+There are a number of pain points with `Copy` types that the lang team is
+interested in exploring, though active experimentation is not currently ongoing.
+
+Some key problems are:
+
+## `Copy` cannot be implemented with non-`Copy` members
+
+There are standard library types where the lack of a `Copy` impl is an
+active pain point, e.g., [`MaybeUninit`](https://github.com/rust-lang/rust/issues/62835)
+and [`UnsafeCell`](https://github.com/rust-lang/rust/issues/25053)
+
+### History
+
+ * `unsafe impl Copy for T` which avoids the requirement that T is recursively
+   Copy, but is obviously unsafe.
+    * https://github.com/rust-lang/rust/issues/25053#issuecomment-218610508
+ * `Copy` is dangerous on types like `UnsafeCell` where `&UnsafeCell<T>`
+   otherwise would not permit access to `T` in [safe
+   code](https://github.com/rust-lang/rust/issues/25053#issuecomment-98447164).
+
+## `Copy` types can be (unintentionally) copied
+
+Even if a type is Copy (e.g., `[u8; 1024]`) it may not be a good idea to make
+use of that in practice, since copying large amounts of data is slow.
+
+This also comes up with `Copy` impls on `Range`, which would generally be
+desirable but is error-prone given the `Iterator/IntoIterator` impls on ranges.
```

Comment:
I think it'd be useful to have canonical examples here of the kinds of code that can be surprising. I think one particularly compelling case has to do with closures:

```rust
let mut x = (0..10);
let c = move || x.next();
println!("{}", x.next()); // prints 0
println!("{}", c()); // prints 0
```

Of course, if you remove the `move` then this code behaves quite differently. (You get 0 and 1.)

You can make a very similar example with `Cell`.

I'm not sure of the best examples otherwise. Certainly if you write `let y = x` you get the same behavior, but it doesn't seem especially *surprising* to me (it's kind of the same as with any integer, for example). 

One interesting example might be `let x = &Cell::new(10)`.